### PR TITLE
add p-defer typings

### PR DESCRIPTION
**Pull request by SamVerschueren**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [ ] Run `npm run lint -- package-name` if a `tslint.json` is present.

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Run `tsc` without errors.
- [x] Include the required [files](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) and header. Base these on the README, *not* on an existing project.

Adding definitions for [p-defer](https://github.com/sindresorhus/p-defer).

### CODE REVIEW

File: p-defer/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "target": "es6",
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictNullChecks": false,
```
Comment (andy-ms) [inline p-defer/tsconfig.json]: Why not?

Comment (SamVerschueren) [inline p-defer/tsconfig.json]: Good question. I created a Yeoman generator ([generator-dt](https://github.com/SamVerschueren/generator-dt)) that scaffolds out most of the files. It looks like I picked a `tslint` without strict null checks as the base of my generator. I'll remove it.

File: p-defer/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+// Definitions by: Sam Verschueren <https://github.com/SamVerschueren>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+interface DeferredPromise<T> {
+	resolve<U>(value: U | PromiseLike<U>): Promise<U>;
+	reject(reason: any): Promise<never>;
+	promise: Promise<T>;
+}
+
+declare const pDefer: <T>() => DeferredPromise<T>;
```
Comment (andy-ms) [inline p-defer/index.d.ts]: Prefer `declare function`

File: p-defer/p-defer-tests.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import pDefer = require('p-defer');
+
+function delay(ms: number) {
+    const deferred = pDefer();
+    setTimeout(deferred.resolve, ms, '🦄');
+    return deferred.promise;
+}
+
+delay(100).then(console.log);
```
Comment (andy-ms) [inline p-defer/p-defer-tests.ts]: `console.log` doesn't test the type of its argument. Try this:
```ts
let s: string;
async function f() { s = await delay(100); }
```

File: p-defer/tsconfig.json (hunk 1 above)
Comment (andy-ms) [inline p-defer/tsconfig.json]: We have a `npm run new-package` script for that.

Comment (SamVerschueren) [inline p-defer/tsconfig.json]: I missed that command, though my generator fills in the header entirely and allows me to extend it in the future. It's just for myself though...

**Pull request merged**